bin/pgenv

```diff
@@ -705,40 +722,85 @@ case $1 in
+            # try to get the most compressed archive
+            if [[ $v =~ ^[1-7]\. ]]; then
+                PG_TARBALL="postgresql-$v.tar.gz"
+                TAR_OPTS="zxf"
+            else
+                PG_TARBALL="postgresql-$v.tar.bz2"
+                TAR_OPTS="jxf"
+            fi
+
+            if [ ! -f $PG_TARBALL ]; then
+                $PGENV_CURL -fLO ${PGENV_DOWNLOAD_ROOT}/v$v/${PG_TARBALL}
+            fi
+
+
+            # warn if no configuration was loaded
+            if [ -z "$PGENV_CONFIGURATION_FILE" ]; then
+                echo "WARNING: no configuration file found for version $v"
+                echo "HINT: if you wish to customize the build process please"
+                echo "stop the execution within 5 seconds (CTRL-c) and run "
+                echo "    pgenv config write $v && pgenv config edit $v"
+                echo "adjust 'configure' and 'make' options and flags and run again"
+                echo "    pgenv build $v"
+                echo
+                sleep 5
+            fi
+
+
+            # Unpack the source.
+            rm -rf "postgresql-$v"
+            $PGENV_TAR $TAR_OPTS $PG_TARBALL
+            cd postgresql-$v
+
+
         else
-            PG_TARBALL="postgresql-$v.tar.bz2"
-            TAR_OPTS="jxf"
-        fi
+            # here there is the build from git for 'dev'
+            BUILD_DIR=postgresql-dev
 
-        if [ ! -f $PG_TARBALL ]; then
-            $PGENV_CURL -fLO ${PGENV_DOWNLOAD_ROOT}/v$v/${PG_TARBALL}
-        fi
+            # initialize variables if not explicitly set
+            if [ -z "$PGENV_GIT_REPO" ]; then
+                PGENV_GIT_REPO="https://git.postgresql.org/git/postgresql.git"
+            fi
 
+            # if the user wants a specific branch, build the
+            # command line options
+            if [ ! -z "$PGENV_GIT_CHECKOUT_BRANCH" ]; then
+                PGENV_GIT_CHECKOUT="-b $PGENV_GIT_CHECKOUT_BRANCH --track origin/$PGENV_GIT_CHECKOUT_BRANCH"
+            else
+                # avoid tainting
+                PGENV_GIT_CHECKOUT=""
+            fi
 
-        # warn if no configuration was loaded
-        if [ -z "$PGENV_CONFIGURATION_FILE" ]; then
-            echo "WARNING: no configuration file found for version $v"
-            echo "HINT: if you wish to customize the build process please"
-            echo "stop the execution within 5 seconds (CTRL-c) and run "
-            echo "    pgenv config write $v && pgenv config edit $v"
-            echo "adjust 'configure' and 'make' options and flags and run again"
-            echo "    pgenv build $v"
-            echo
-            sleep 5
+            pgenv_debug "Checking out [$PGENV_GIT_REPO] @ [$PGENV_GIT_CHECKOUT]"
+
+            # download the sources
+            if [ -d "$BUILD_DIR" ]; then
+                # directory already existing, fetch new sources
+                cd "$BUILD_DIR"
+                git fetch $PGENV_GIT_REPO
```

Comment:
This won't pull into the current branch. I think you want to also `git pull`.

Reply:
I'm thinking we should take the git "update" part out of the process. In my initial idea there was the capability to checkout a branche and build such branch, but pulling requires a merging, so it can lead to a mess if you want to build a dev version in the past (e.g., branch `REL_10_STABLE`) after your working tree is advanced. If we leave only the repository cloning (first time) and the build cycle, the user is free to checkout and move the git head to whatever he wants and start the build. What do you think?

Reply:
Then there is no need to fetch.

Reply:
That's not fully clear to me. At the moment, commit 3d9d12a, the repository is cloned if it seems there's no local copy of the source tree (i.e., `src/postgresql-dev` directory does not exist) or is pulled if appears to be a local copy. 
I was thinking about removing the pull part and leave only the initial clone, to give the user more power to `build-dev` against a modified version of the source tree.